Remove JS Scraper, Revamp DB Structure

dog-broad opened the pull request:
Also reformatted code to IntelliJ WebStorm Standards

dog-broad committed:
Remove JS Scraper, Revamp DB Structure

Also reformatted code to IntelliJ WebStorm Standards

dog-broad commented:
@CodiumAI-Agent /describe

dog-broad committed:
Integrate ProfileForm with backend API and add user data verification

The ProfileForm component has been updated to integrate with the backend API for fetching and updating user profile data. This commit also introduces a new feature to verify user data before saving.

Changes:
- Added state variable `userDataEmpty` to track whether user data is empty.
- Updated fetchData function to handle cases where user data is empty or not found.
- Added logic to verify all usernames before saving data.
- Implemented a dirty hack to trigger a re-render after successful data saving to update the UI.

Additionally, the handleVerify function now properly handles different verification endpoints for different platforms.

Fixes:
- Resolved a bug where the application would crash if the fetched user data was empty or not found.
- Fixed an issue where the application did not properly handle empty or missing usernames during data saving.

dog-broad commented:
The schema of the `supabase` DataBase during this commit is as following:

<details><summary> Schema </summary><p>

```sql
-- Drop existing tables if they exist
DROP TABLE IF EXISTS platforms;
DROP TABLE IF EXISTS users;

-- Create users table
CREATE TABLE users (
    user_id VARCHAR PRIMARY KEY, -- UUID can be replaced with VARCHAR
    name VARCHAR,
    email VARCHAR,
    photo_url VARCHAR,
    year_of_passing INT,
    hall_ticket_no VARCHAR UNIQUE, -- Ensure hall_ticket_no is unique
    updated_at TIMESTAMP,
    platforms JSONB -- Store platforms data as JSONB
);

-- Create platforms table
CREATE TABLE platforms (
    id SERIAL PRIMARY KEY,
    user_id VARCHAR REFERENCES users(user_id) ON DELETE CASCADE,
    platform_name VARCHAR(255),
    platform_username VARCHAR(255),
    verification_status BOOLEAN,
    score INTEGER,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

-- Add a unique constraint to ensure no duplicates for each platform_name
ALTER TABLE platforms ADD CONSTRAINT unique_platform_username_per_platform UNIQUE (platform_name, platform_username);

-- Create an index on user_id for efficient querying
CREATE INDEX idx_user_id ON platforms(user_id);

-- Create a function to handle updates and inserts
CREATE OR REPLACE FUNCTION update_platforms()
RETURNS TRIGGER AS $$
DECLARE
    platform_data JSONB;
    v_status BOOLEAN;
BEGIN
    -- Loop through each platform in the JSONB data
    FOR platform_data IN 
        SELECT jsonb_each_text(NEW.platforms) 
    LOOP
        -- Parse the verification status as boolean
        v_status := (platform_data.value ->> 'verification_status')::BOOLEAN;
        
        -- Check if the platform exists in the platforms table
        IF EXISTS (
            SELECT 1 
            FROM platforms 
            WHERE user_id = NEW.user_id 
            AND platform_name = platform_data.key
        ) THEN
            -- If exists, update the existing row
            UPDATE platforms
            SET platform_username = (platform_data.value ->> 'platform_username'),
                verification_status = v_status,
                updated_at = NEW.updated_at
            WHERE user_id = NEW.user_id 
            AND platform_name = platform_data.key;
        ELSE
            -- If not exists, insert a new row
            INSERT INTO platforms (user_id, platform_name, platform_username, verification_status, score, created_at, updated_at)
            VALUES (
                NEW.user_id, 
                platform_data.key,
                (platform_data.value ->> 'platform_username'),
                v_status,
                NULL, -- You can set the score to NULL initially
                NEW.updated_at, 
                NEW.updated_at
            );
        END IF;
    END LOOP;
    
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

-- Create triggers for insert and update operations on the users table
CREATE TRIGGER users_insert_trigger
AFTER INSERT ON users
FOR EACH ROW
EXECUTE FUNCTION update_platforms();

CREATE TRIGGER users_update_trigger
AFTER UPDATE ON users
FOR EACH ROW
EXECUTE FUNCTION update_platforms();
```

</p></details>

But this schema has however forced me into errors such as(Fetched from supabase Postgres logs):

```json
{
  "file": null,
  "host": "db-msyamhipiqrodsygmeyd",
  "metadata": [],
  "parsed": [
    {
      "application_name": "postgrest",
      "backend_type": "client backend",
      "command_tag": "SELECT",
      "connection_from": "::1:56690",
      "context": "JSON data, line 1: (...\nPL/pgSQL function update_platforms() line 7 at FOR over SELECT rows",
      "database_name": "postgres",
      "detail": "Token \"(\" is invalid.",
      "error_severity": "ERROR",
      "hint": null,
      "internal_query": null,
      "internal_query_pos": null,
      "leader_pid": null,
      "location": null,
      "process_id": 39108,
      "query": "WITH pgrst_source AS (INSERT INTO \"public\".\"users\"(\"email\", \"hall_ticket_no\", \"name\", \"photo_url\", \"platforms\", \"updated_at\", \"user_id\", \"year_of_passing\") SELECT \"pgrst_body\".\"email\", \"pgrst_body\".\"hall_ticket_no\", \"pgrst_body\".\"name\", \"pgrst_body\".\"photo_url\", \"pgrst_body\".\"platforms\", \"pgrst_body\".\"updated_at\", \"pgrst_body\".\"user_id\", \"pgrst_body\".\"year_of_passing\" FROM (SELECT $1 AS json_data) pgrst_payload, LATERAL (SELECT CASE WHEN json_typeof(pgrst_payload.json_data) = 'array' THEN pgrst_payload.json_data ELSE json_build_array(pgrst_payload.json_data) END AS val) pgrst_uniform_json, LATERAL (SELECT \"email\", \"hall_ticket_no\", \"name\", \"photo_url\", \"platforms\", \"updated_at\", \"user_id\", \"year_of_passing\" FROM json_to_recordset(pgrst_uniform_json.val) AS _(\"email\" character varying, \"hall_ticket_no\" character varying, \"name\" character varying, \"photo_url\" character varying, \"platforms\" jsonb, \"updated_at\" timestamp without time zone, \"user_id\" character varying, \"year_of_passing\" integer) ) pgrst_body WHERE set_config('pgrst.inserted', (coalesce(nullif(current_setting('pgrst.inserted', true), '')::int, 0) + 1)::text, true) <> '0' ON CONFLICT(\"user_id\") DO UPDATE SET \"email\" = EXCLUDED.\"email\", \"hall_ticket_no\" = EXCLUDED.\"hall_ticket_no\", \"name\" = EXCLUDED.\"name\", \"photo_url\" = EXCLUDED.\"photo_url\", \"platforms\" = EXCLUDED.\"platforms\", \"updated_at\" = EXCLUDED.\"updated_at\", \"user_id\" = EXCLUDED.\"user_id\", \"year_of_passing\" = EXCLUDED.\"year_of_passing\"WHERE set_config('pgrst.inserted', (coalesce(nullif(current_setting('pgrst.inserted', true), '')::int, 0) - 1)::text, true) <> '-1' RETURNING 1) SELECT '' AS total_result_set, pg_catalog.count(_postgrest_t) AS page_total, array[]::text[] AS header, ''::text AS body, nullif(current_setting('response.headers', true), '') AS response_headers, nullif(current_setting('response.status', true), '') AS response_status, nullif(current_setting('pgrst.inserted', true),'')::int AS response_inserted FROM (SELECT * FROM pgrst_source) _postgrest_t",
      "query_id": -1747951979318262500,
      "query_pos": null,
      "session_id": "6642f62b.98c4",
      "session_line_num": 3,
      "session_start_time": "2024-05-14 05:27:07 UTC",
      "sql_state_code": "22P02",
      "timestamp": "2024-05-14 05:27:28.669 UTC",
      "transaction_id": 2150,
      "user_name": "authenticator",
      "virtual_transaction_id": "6/63958"
    }
  ],
  "parsed_from": null,
  "project": null,
  "source_type": null
}
```

I sadly couldn't figure out a proper fix for this

dog-broad commented:
And finally, I decided to re-write the database schema again with new `platform_data` table as I have been facing issue in Django Backend too

The new schema is as following:

```sql
-- Drop existing tables if they exist
DROP TABLE IF EXISTS platform_data;
DROP TABLE IF EXISTS users;

-- Create users table
CREATE TABLE users (
    user_id VARCHAR PRIMARY KEY, -- UUID can be replaced with VARCHAR
    name VARCHAR,
    email VARCHAR,
    photo_url VARCHAR,
    year_of_passing INT,
    hall_ticket_no VARCHAR UNIQUE, -- Ensure hall_ticket_no is unique
    updated_at TIMESTAMP,
    platforms JSONB -- Store platforms data as JSONB
);

-- Create platform_data table
CREATE TABLE platform_data (
    platform_id SERIAL PRIMARY KEY,
    user_id VARCHAR REFERENCES users(user_id),
    platform_name VARCHAR NOT NULL,
    platform_username VARCHAR NOT NULL,
    verification_status VARCHAR NOT NULL,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
    score INT,
    UNIQUE(platform_name, platform_username)
);

-- Create a trigger function to populate platform_data
CREATE OR REPLACE FUNCTION populate_platform_data()
RETURNS TRIGGER AS $$
DECLARE
    platform RECORD;
BEGIN
    -- Loop through each platform in the JSONB data
    FOR platform IN SELECT * FROM jsonb_each(NEW.platforms)
    LOOP
        -- Try to update the existing platform data
        UPDATE platform_data
        SET 
            platform_username = platform.value->>'platform_username',
            verification_status = platform.value->>'verification_status',
            updated_at = CURRENT_TIMESTAMP,
            score = CASE 
                        WHEN (platform.value->>'score') IS NOT NULL THEN (platform.value->>'score')::int 
                        ELSE NULL 
                    END
        WHERE 
            user_id = NEW.user_id
            AND platform_name = platform.key::text;
        
        -- Check if the row was updated
        IF NOT FOUND THEN
            -- If not, insert a new row
            INSERT INTO platform_data (user_id, platform_name, platform_username, verification_status, created_at, updated_at, score)
            VALUES (
                NEW.user_id,
                platform.key::text,
                platform.value->>'platform_username',
                platform.value->>'verification_status',
                CURRENT_TIMESTAMP,
                CURRENT_TIMESTAMP,
                CASE 
                    WHEN (platform.value->>'score') IS NOT NULL THEN (platform.value->>'score')::int 
                    ELSE NULL 
                END
            );
        END IF;
    END LOOP;

    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

-- Create a trigger to call the function after insert or update on users table
CREATE TRIGGER after_users_insert_or_update
AFTER INSERT OR UPDATE ON users
FOR EACH ROW
EXECUTE FUNCTION populate_platform_data();

```

dog-broad commented:
@CodiumAI-Agent /review

QodoAI-Agent commented:
## PR Review 🔍

<table>
<tr>
<tr><td> ⏱️&nbsp;<strong>Estimated&nbsp;effort&nbsp;to&nbsp;review [1-5]</strong></td><td>

3, because the PR includes a significant number of changes across multiple files, primarily focused on code formatting and minor logic adjustments. The changes are spread across both front-end React components and backend service configurations, requiring a thorough review to ensure functionality remains intact and styling conforms to the new standards.


</td></tr>
<tr><td> 🧪&nbsp;<strong>Relevant tests</strong></td><td>

No


</td></tr>
<tr><td> ⚡&nbsp;<strong>Possible issues</strong></td><td>

<strong>Possible Bug:</strong> The use of `sessionStorage` for storing user data in `MainPage.js` and `ProfileForm.js` might not persist the user's data across multiple tabs or after the browser is closed, which could lead to data loss or inconsistency.

</td></tr>
<tr><td> 🔒&nbsp;<strong>Security concerns</strong></td><td>

No

</td></tr>
</table>


<details><summary> <strong>Code feedback:</strong></summary>

<hr><table><tr><td>relevant file</td><td>src/components/MainPage.js
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Consider using `localStorage` instead of `sessionStorage` for storing user data to ensure data persistence across browser sessions. This change will help maintain user data even after the browser is closed, providing a better user experience. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/dog-broad/leaderboard-scraper/pull/1/files#diff-e7a268f75db50a1df3bc2f5729e7b6e04cadaa4b40b409115f5720986f6b247aR25'>const cachedUserData = sessionStorage.getItem('userData');</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>src/components/ProfileForm.js
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Implement form validation for the `yearOfPassing` input to ensure it is a future year and not just any number greater than 2025. This will prevent users from entering invalid years that could lead to data inconsistencies. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/dog-broad/leaderboard-scraper/pull/1/files#diff-1de78120428903277288f88b459cf1c8587d1eba75c3d3731c5a9ab7c7245a8cR229'>if (isNaN(selectedYear) || selectedYear < 2025) {</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>src/components/ScoresPage.js
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Add error handling for the `fetchUserData` function to manage cases where data fetching might fail. This will improve the robustness of your application by handling potential errors gracefully. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/dog-broad/leaderboard-scraper/pull/1/files#diff-a893a9ad679bf094e8c567aec1e4bac5bfe0d99469cf98ef8507b16f2e475f92R26'>fetchUserData().then(r => r);</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>src/components/UserDataPage.js
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Optimize the search functionality by debouncing the input handler. This prevents the app from making excessive computations or re-renders as the user types, enhancing performance and user experience. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/dog-broad/leaderboard-scraper/pull/1/files#diff-0e61cc49e359f0a245ae305d7a0ebc0a2a05d78547ce95ebfe53bd78f2a283ebR36'>const handleSearch = (e) => {</a></td></tr></table><hr>

</details>

dog-broad commented:
> And finally, I decided to re-write the database schema again with new `platform_data` table as I have been facing issue in Django Backend too
> 
> The new schema is as following:
> 
> ```sql
> -- Drop existing tables if they exist
> DROP TABLE IF EXISTS platform_data;
> DROP TABLE IF EXISTS users;
> 
> -- Create users table
> CREATE TABLE users (
>     user_id VARCHAR PRIMARY KEY, -- UUID can be replaced with VARCHAR
>     name VARCHAR,
>     email VARCHAR,
>     photo_url VARCHAR,
>     year_of_passing INT,
>     hall_ticket_no VARCHAR UNIQUE, -- Ensure hall_ticket_no is unique
>     updated_at TIMESTAMP,
>     platforms JSONB -- Store platforms data as JSONB
> );
> 
> -- Create platform_data table
> CREATE TABLE platform_data (
>     platform_id SERIAL PRIMARY KEY,
>     user_id VARCHAR REFERENCES users(user_id),
>     platform_name VARCHAR NOT NULL,
>     platform_username VARCHAR NOT NULL,
>     verification_status VARCHAR NOT NULL,
>     created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
>     updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
>     score INT,
>     UNIQUE(platform_name, platform_username)
> );
> 
> -- Create a trigger function to populate platform_data
> CREATE OR REPLACE FUNCTION populate_platform_data()
> RETURNS TRIGGER AS $$
> DECLARE
>     platform RECORD;
> BEGIN
>     -- Loop through each platform in the JSONB data
>     FOR platform IN SELECT * FROM jsonb_each(NEW.platforms)
>     LOOP
>         -- Try to update the existing platform data
>         UPDATE platform_data
>         SET 
>             platform_username = platform.value->>'platform_username',
>             verification_status = platform.value->>'verification_status',
>             updated_at = CURRENT_TIMESTAMP,
>             score = CASE 
>                         WHEN (platform.value->>'score') IS NOT NULL THEN (platform.value->>'score')::int 
>                         ELSE NULL 
>                     END
>         WHERE 
>             user_id = NEW.user_id
>             AND platform_name = platform.key::text;
>         
>         -- Check if the row was updated
>         IF NOT FOUND THEN
>             -- If not, insert a new row
>             INSERT INTO platform_data (user_id, platform_name, platform_username, verification_status, created_at, updated_at, score)
>             VALUES (
>                 NEW.user_id,
>                 platform.key::text,
>                 platform.value->>'platform_username',
>                 platform.value->>'verification_status',
>                 CURRENT_TIMESTAMP,
>                 CURRENT_TIMESTAMP,
>                 CASE 
>                     WHEN (platform.value->>'score') IS NOT NULL THEN (platform.value->>'score')::int 
>                     ELSE NULL 
>                 END
>             );
>         END IF;
>     END LOOP;
> 
>     RETURN NEW;
> END;
> $$ LANGUAGE plpgsql;
> 
> -- Create a trigger to call the function after insert or update on users table
> CREATE TRIGGER after_users_insert_or_update
> AFTER INSERT OR UPDATE ON users
> FOR EACH ROW
> EXECUTE FUNCTION populate_platform_data();
> ```

https://github.com/dog-broad/leaderboard-scraper/commit/4de17cc8f52f51789fccce997c5d1f36b858167f

The above commit integrates with the new schema provided

dog-broad committed:
Refactor scores fetching and update logic

- Updated API endpoint from `/platforms` to `/platform_data` for fetching user data
- Added a call to a Django API to update scores with a POST request
- Updated the display logic to show the correct score value

This commit refactors the scores fetching and update logic to improve functionality and accuracy.
